src/django_components/extensions/url.py

```diff
@@ -0,0 +1,180 @@
+
+
+# NOTE: `WeakKeyDictionary` is NOT a generic pre-3.9
+if sys.version_info >= (3, 9):
+    ComponentRouteCache = WeakKeyDictionary[Type["Component"], URLRoute]
+else:
+    ComponentRouteCache = WeakKeyDictionary
+
+
+def _get_component_route_name(component: Union[Type["Component"], "Component"]) -> str:
+    return f"__component_url__{component._class_hash}"
+
+
+def get_component_url(component: Union[Type["Component"], "Component"]) -> str:
+    """
+    Get the URL for a [`Component`](../api#django_components.Component).
+
+    Raises `RuntimeError` if the component is not public.
+
+    Read more about [Component views and URLs](../../concepts/fundamentals/component_views_urls).
+
+    **Example:**
+
+    ```py
+    from django_components import Component, get_component_url
+
+    class MyComponent(Component):
+        class Url:
+            public = True
+
+    # Get the URL for the component
+    url = get_component_url(MyComponent)
+    ```
+    """
+    url_cls: Optional[Type[ComponentUrl]] = getattr(component, "Url", None)
+    if url_cls is None or not url_cls.public:
+        raise RuntimeError("Component URL is not available - Component is not public")
+
+    route_name = _get_component_route_name(component)
+    return django.urls.reverse(route_name)
+
+
+class ComponentUrl(ComponentExtension.ExtensionClass):  # type: ignore
+    """
+    The interface for `Component.Url`.
+
+    This class is used to configure whether the component should be available via a URL.
+
+    Read more about [Component views and URLs](../../concepts/fundamentals/component_views_urls).
+
+    **Example:**
+    ```python
+    from django_components import Component
+
+    class MyComponent(Component):
+        class Url:
+            public = True
+
+    # Get the URL for the component
+    url = get_component_url(MyComponent)
+    ```
+    """
+
+    public: bool = False
+    """
+    Whether this [`Component`](../api#django_components.Component) should be available
+    via a URL. Defaults to `False`.
+
+    If `True`, the Component will have its own unique URL path.
+
+    You can use this to write components that will correspond to HTML fragments
+    for HTMX or similar libraries.
+
+    To obtain the component URL, either access the url from
+    [`Component.Url.url`](../api#django_components.ComponentUrl.url) or
+    use the [`get_component_url()`](../api#django_components.get_component_url) function.
+
+    **Example:**
+
+    ```py
+    from django_components import Component, get_component_url
+
+    class MyComponent(Component):
+        class Url:
+            public = True
+
+    # Get the URL for the component
+    url = get_component_url(MyComponent)
+    ```
+    """
+
+    @property
+    def url(self) -> str:
+        """
+        The URL for the component.
+
+        Raises `RuntimeError` if the component is not public.
+        """
+        return get_component_url(self.component.__class__)
+
+
+class UrlExtension(ComponentExtension):
```

Comment:
The main logic of this PR - the extension listens to when a component class is created or deleted, and creates / deletes a URL route that has the `Component.as_view()`. as the handler.